File: 0077-combinations/0077-combinations.java
class Solution {
    static List<List<Integer>> answer;
    
    public List<List<Integer>> combine(int n, int k) {
        answer = new ArrayList<>();
        combi(1, n, k, new ArrayList<>());
        return answer;
    }
    
    private void combi(int start, int n, int k, List<Integer> list) {
        if (list.size() == k) {
            answer.add(new ArrayList<>(list));
            return;
        }
        
        for (int i = start; i <= n; i++) {
            list.add(i);
            combi(i + 1, n, k, list);
            list.remove(Integer.valueOf(i));
        }
        
    }
}

File: 0077-combinations/0077-combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let answer = [];
    com(0, n, k, 1, answer, []);
    return answer;
};

const com = (level, n, k, start, answer, tmp) => {
    if (level === k) {
        answer.push(tmp.slice());  // tmp 배열의 복사본을 추가
        return;
    }

    for (let i = start; i <= n; i++) {
        tmp.push(i);
        com(level + 1, n, k, i + 1, answer, tmp);
        tmp.pop();
    }
}
